### ISSUE TITLE
Code execution stalls if there's too much stdout

### ISSUE THREAD

**Issue by mortenpi** [fredrikekre/Literate.jl#138]
Literate stalls when executing code that outputs a lot of text into stdout (also stderr probably). My best guess is that this is due to how we do the output capturing and some buffer gets full somewhere. FWIW, there is a good chance that some of the reports of Documenter stalling are due to the same issue.

For an MWE, running the following Literate file

```julia
for _ in 1:20
    println("="^80)
end

# This should stall?

for _ in 1:2000
    println("="^80)
end

# And this would complete again?

println("="^80)
```

with `using Literate; Literate.notebook("literate-mwe.jl", execute=true)` stalls when executing the second block. The stacktrace is always in a `wait` call of the I/O:

```
ERROR: InterruptException:
Stacktrace:
 [1] poptask(::Base.InvasiveLinkedListSynchronized{Task}) at ./task.jl:704
 [2] wait at ./task.jl:712 [inlined]
 [3] uv_write(::Base.PipeEndpoint, ::Ptr{UInt8}, ::UInt64) at ./stream.jl:933
 [4] unsafe_write(::Base.PipeEndpoint, ::Ptr{UInt8}, ::UInt64) at ./stream.jl:1005
 [5] unsafe_write at ./io.jl:622 [inlined]
 [6] unsafe_write(::Base.PipeEndpoint, ::Base.RefValue{UInt8}, ::Int64) at ./io.jl:620
 [7] write at ./io.jl:623 [inlined]
 [8] write(::Base.PipeEndpoint, ::UInt8) at ./stream.jl:1043
 [9] write at ./io.jl:676 [inlined]
 [10] print at ./char.jl:229 [inlined]
 [11] println(::Base.PipeEndpoint) at ./coreio.jl:5
 [12] println() at ./coreio.jl:4
 [13] (::Literate.Documenter.var"#1#2"{Literate.var"#38#39"{Module,String},Base.TTY,Base.TTY,Pipe,Array{UInt8,1}})() at /home/mortenpi/Projects/highZ/dev/Literate/src/Documenter.jl:30
 [14] with_logstate(::Function, ::Any) at ./logging.jl:408
 [15] with_logger at ./logging.jl:514 [inlined]
 [16] withoutput at /home/mortenpi/Projects/highZ/dev/Literate/src/Documenter.jl:23 [inlined]
 [17] execute_block(::Module, ::String) at /home/mortenpi/Projects/highZ/dev/Literate/src/Literate.jl:720
 [18] execute_notebook(::Dict{Any,Any}) at /home/mortenpi/Projects/highZ/dev/Literate/src/Literate.jl:625
 [19] (::Literate.var"#34#36")() at /home/mortenpi/Projects/highZ/dev/Literate/src/Literate.jl:606
 [20] cd(::Literate.var"#34#36", ::String) at ./file.jl:104
 [21] jupyter_notebook(::Array{Literate.Chunk,1}, ::Dict{String,Any}) at /home/mortenpi/Projects/highZ/dev/Literate/src/Literate.jl:605
 [22] notebook(::String, ::String; config::Dict{Any,Any}, kwargs::Base.Iterators.Pairs{Symbol,Bool,Tuple{Symbol},NamedTuple{(:execute,),Tuple{Bool}}}) at /home/mortenpi/Projects/highZ/dev/Literate/src/Literate.jl:542
 [23] top-level scope at REPL[1]:1
```

Tried both Julia 1.5.3 and 1.7.0-dev.213.

**Comment by KristofferC** [comment]
I think you need to have a task that continuously reads from the pipe.

**Comment by KristofferC** [comment]
For example:

```
function withoutput(f)
    # Save the default output streams.
    default_stdout = stdout
    default_stderr = stderr

    # Redirect both the `stdout` and `stderr` streams to a single `Pipe` object.
    pipe = Pipe()
    Base.link_pipe!(pipe; reader_supports_async = true, writer_supports_async = true)
    redirect_stdout(pipe.in)
    redirect_stderr(pipe.in)
    # Also redirect logging stream to the same pipe
    logger = Base.CoreLogging.SimpleLogger(pipe.in)

    # Bytes written to the `pipe` are captured in `output` and converted to a `String`.
    output = UInt8[]

    io = IOBuffer()

    t = @async while !eof(pipe)
        write(io, readavailable(pipe))
    end

    # Run the function `f`, capturing all output that it might have generated.
    # Success signals whether the function `f` did or did not throw an exception.
    result, success, backtrace = Base.CoreLogging.with_logger(logger) do
        try
            f(), true, Vector{Ptr{Cvoid}}()
        catch err
            err, false, catch_backtrace()
        finally
            # Force at least a single write to `pipe`, otherwise `readavailable` blocks.
            println()
            # Restore the original output streams.
            redirect_stdout(default_stdout)
            redirect_stderr(default_stderr)
            close(pipe)
            wait(t)
        end
    end
    return result, success, backtrace, chomp(String(take!(io)))
end
```

**Comment by francispoulin** [comment]
Silly question but how do you se this function in the context of the following lines?

```
using Literate; 
Literate.notebook("literate-mwe.jl", execute=true)
```

**Comment by fredrikekre** [comment]
I have not seen this before, did something change in Julia? Does it stall on e.g. 1.5 too?

In any case, is Kristoffers suggestion something that can/should be implemented in `IOCapture` and thus fixed here by merging  #124 @mortenpi? Or perhaps that package does not suffer from this? I believe it should be mostly equivalent code though.

**Comment by mortenpi** [comment]
Yep, as expected, IOCapture suffers from the same issue. Also, I can also replicate it on Julia 1.0, 1.2 etc. Not exactly sure why it hasn't cropped up before, but it does need something like 40KiB of bytes written to stdout, which is way more than a user usually outputs from a block I reckon. In my case, I managed to run into this because I added some warning printouts to a function that was being called in a loop.

I think getting the fix in via IOCapture is a good idea. #124 just needs a version bump once a we tag a version with the fix.
